src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java

```diff
@@ -19,100 +21,174 @@
+    }
+
+    @Test
+    void givenMultipleListeners_eachShouldReceiveSeparateEvents() throws Exception {
+        MongoCDCManager manager = new MongoCDCManager(mongoConfig);
+
+        TestChangeStreamListener listener0 = new TestChangeStreamListener();
+        TestChangeStreamListener listener1 = new TestChangeStreamListener();
+        TestChangeStreamListener listener2 = new TestChangeStreamListener();
+        manager.registerListener(listener0, List.of(0));
+        manager.registerListener(listener1, List.of(1));
+        manager.registerListener(listener2, List.of(2));
+        manager.start();
+
+        insertDocumentsToAllPartitions();
+
+        // Wait for CDC event to arrive
+        Thread.sleep(500);
+
+        List<ChangeStreamDocument<Document>> events0 = listener0.getEvents();
+        assertEquals(1, events0.size());
+        assertEquals(OperationType.INSERT, events0.get(0).getOperationType());
+        assertEquals(0, events0.get(0).getFullDocument().getInteger("testValue"));
+
+        List<ChangeStreamDocument<Document>> events1 = listener1.getEvents();
+        assertEquals(1, events1.size());
+        assertEquals(OperationType.INSERT, events1.get(0).getOperationType());
+        assertEquals(1, events1.get(0).getFullDocument().getInteger("testValue"));
+
+        List<ChangeStreamDocument<Document>> events2 = listener2.getEvents();
+        assertEquals(1, events2.size());
+        assertEquals(OperationType.INSERT, events2.get(0).getOperationType());
+        assertEquals(2, events2.get(0).getFullDocument().getInteger("testValue"));
+    }
+
+    @Test
+    void shouldNotHandleEventsWhenListenerIsDeregisteredFromAllPartitions() throws Exception {
+        MongoCDCManager manager = new MongoCDCManager(mongoConfig);
+
+        TestChangeStreamListener listener0 = new TestChangeStreamListener();
+        manager.registerListenerToAllPartitions(listener0);
+        manager.start();
+
+        insertDocumentsToAllPartitions();
+
+        // Wait for CDC event to arrive
+        Thread.sleep(500);
+
+        List<ChangeStreamDocument<Document>> events0 = listener0.getEvents();
+        assertEquals(3, events0.size());
+
+        // when
+        manager.deregisterListenerFromAllPartitions(listener0);
+
+        // then
+        var deleteResult0 = collection.deleteOne(Filters.eq("_id", new ObjectId(TestIds.MOD_0_ID)));
+        var deleteResult1 = collection.deleteOne(Filters.eq("_id", new ObjectId(TestIds.MOD_1_ID)));
+        var deleteResult2 = collection.deleteOne(Filters.eq("_id", new ObjectId(TestIds.MOD_2_ID)));
+        assertEquals(1, deleteResult0.getDeletedCount());
+        assertEquals(1, deleteResult1.getDeletedCount());
+        assertEquals(1, deleteResult2.getDeletedCount());
+
+        // Wait for CDC event to arrive
+        Thread.sleep(500);
+
+        // only old events are available
+        List<ChangeStreamDocument<Document>> events1 = listener0.getEvents();
+        assertEquals(3, events1.size());
```

Comment:
3 should be replaced here by events0.size()

Reply:
done